client/session.go

```diff
@@ -1437,20 +1437,27 @@ func (s *session) streamBlocksBatchFromPeer(
 			continue
 		}
 
+		missed := 0
 		for j := range result.Elements[i].Blocks {
 			if j >= len(batch[i].blocks) {
 				s.log.Errorf("stream blocks response from peer %s returned more blocks than expected", peer.Host().String())
 				break
 			}
 
-			block := result.Elements[i].Blocks[j]
+			// Index of the received block could be offset by missed blocks
+			block := result.Elements[i].Blocks[j-missed]
+
 			if block.Start != batch[i].blocks[j].start.UnixNano() {
+				missed++
```

Comment:
To avoid this we always order the timestamps ascending, both on the callers and the callees end and the callee only ever returns a subset of timestamps so the only case to cover here is missing blocks.

These blocks can be pretty big considering this is a batch call of 128 IDs so I'd prefer not to allocate a map here if at all possible.